src/passes/GlobalTypeOptimization.cpp

```diff
@@ -175,16 +189,39 @@ struct GlobalTypeOptimization : public Pass {
         vec[i] = true;
       }
 
-      // Process removability. First, see if we can remove anything before we
-      // start to allocate info for that.
-      if (std::any_of(infos.begin(), infos.end(), [&](const FieldInfo& info) {
-            return !info.hasRead;
-          })) {
+      // Process removability. We check separately for the ability to
+      // remove in a general way based on sub+super-propagated info (that is,
+      // fields that are not used in sub- or super-types, and so we can
+      // definitely remove them from all the relevant types) and also in the
+      // specific way that only works for removing at the end, which as
+      // mentioned above only looks at super-types.
+      std::set<Index> removableIndexes;
+      for (Index i = 0; i < fields.size(); i++) {
+        if (!subSuper[i].hasRead) {
+          removableIndexes.insert(i);
+        }
+      }
+      for (int i = int(fields.size()) - 1; i >= 0; i--) {
+        // Unlike above, a write would stop us here: above we propagated to both
+        // sub- and super-types, which means if we see no reads then there is no
+        // possible read of the data at all. But here we just propagated to
+        // subtypes, and so we need to care about the case where the parent
+        // writes to a field but does not read from it - we still need those
+        // writes to happen as children may read them.
```

Comment:
Can't we look at the children and determine precisely whether or not there is a read in a child? I don't see why we need to be conservative here.

Reply:
If children do not read this, then that means nothing reads this at all, and it would be handled in the other case anyhow. I'll add to the comment.